src/core/net/ip6.cpp

```diff
@@ -603,27 +622,240 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+        {
+            // starting the timer for the reassembly timeout
+            mTimer.Start(kIp6ReassemblyTimeoutMilli);
+
+            memcpy(&mFragmentBuffer.mMessageInfo, &aMessageInfo, sizeof(MessageInfo));
+            memcpy(&mFragmentBuffer.mHeader, &aHeader, sizeof(Header));
+
+            mFragmentBuffer.mFragmentIdentification = fragmentHeader.GetIdentification();
+            mFragmentBuffer.mFragmentNextHeader     = fragmentHeader.GetNextHeader();
+
+            VerifyOrExit((mFragmentBuffer.mIp6MsgBuffer = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+            SuccessOrExit(error = mFragmentBuffer.mIp6MsgBuffer->SetLength(aMessage.GetOffset()));
+            SuccessOrExit(error = mFragmentBuffer.mIp6MsgBuffer->SetOffset(aMessage.GetOffset()));
+            // copying the non-fragmentable header to the fragmentation buffer
+            VerifyOrExit(aMessage.CopyTo(0, 0, aMessage.GetOffset(), *mFragmentBuffer.mIp6MsgBuffer) ==
+                             aMessage.GetOffset(),
+                         error = OT_ERROR_ABORT);
+
+            otLogDebgIp6("start reassembly.");
+        }
+
+        otLogInfoIp6("Fragment with id %d received > %d bytes, offset %d", mFragmentBuffer.mFragmentIdentification,
+                     payloadFragment, offset);
+
+        if (offset + payloadFragment + mFragmentBuffer.mIp6MsgBuffer->GetOffset() > kMaxAssembledDatagramLength)
+        {
+            otLogWarnIp6("Package too large for fragment buffer");
+            ExitNow(error = OT_ERROR_NO_BUFS);
+        }
+
+        // increase message buffer if necessary
+        if (mFragmentBuffer.mIp6MsgBuffer->GetLength() <
+            offset + payloadFragment + mFragmentBuffer.mIp6MsgBuffer->GetOffset())
+        {
+            SuccessOrExit(error = mFragmentBuffer.mIp6MsgBuffer->SetLength(offset + payloadFragment +
+                                                                           mFragmentBuffer.mIp6MsgBuffer->GetOffset()));
+        }
+
+        // copy the fragment payload into the message buffer
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + sizeof(fragmentHeader),
+                                     mFragmentBuffer.mIp6MsgBuffer->GetOffset() + offset, payloadFragment,
+                                     *mFragmentBuffer.mIp6MsgBuffer) == (int)payloadFragment,
+                     error = OT_ERROR_ABORT);
+
+        // check if it is the last frame
+        if (fragmentHeader.IsMoreFlagSet() == false)
+        {
+            mFragmentBuffer.mReassemblyPayloadTotal = offset + payloadFragment;
+        }
+
+        if (mFragmentBuffer.mReassemblyPayloadTotal == mFragmentBuffer.mReassemblyPayloadReceived)
+        {
+            // ip6 reassembly complete
+            offset = 0;
+
+            aMessage.SetLength(mFragmentBuffer.mIp6MsgBuffer->GetLength());
+            mFragmentBuffer.mIp6MsgBuffer->CopyTo(0, 0, mFragmentBuffer.mIp6MsgBuffer->GetLength(), aMessage);
```

Comment:
The copy here means we have two copies of the fully reassembled datagram. Would it make more sense to drop the received fragment and call `HandleDatagram()` on the reassembled datagram to avoid the copy?

Reply:
Yes.
Now we drop all fragments and call `HandleDatagram()` with the whole packet in the buffer.